File: python/rise/latex_generation/mystic_spheres/thaumaturgy.py
from rise.latex.mystic_sphere import MysticSphere
from rise.latex.spell import Spell
from rise.latex.effects import Effects


# Primary: utility
# Secondary: debuff
# Tertiary: buff
# None: damage
thaumaturgy = MysticSphere(
    name="Thaumaturgy",
    short_description="Suppress and manipulate magical effects",
    cantrips=[
        Effects(
            "Sense Magic",
            "Yourself",
            """
            You gain a +4 bonus to the Spellsense skill until the end of the next round.
        """,
            scaling="""
            \\rank<2> The bonus increases to +6.
            \\rank<4> The bonus increases to +8.
            \\rank<6> The bonus increases to +10.
        """,
            tags=[],
        ),
        Effects(
            "Sense Magical Potential",
            "One creature within \\rngshort range",
            """
            You discern whether the target has any \\glossterm<magical> abilities.
            This does not give you any information about the nature of those magical abilities.
        """,
            scaling="""
            \\rank<2> You can also discern whether the target has the ability to cast spells of any kind.
            \\rank<4> You can also discern which \\glossterm<magic sources> the target has access to, if they have access to any.
            \\rank<6> You can also discern which \\glossterm<mystic spheres> the target has access to, if they have access to any.
        """,
            tags=[],
        ),
    ],
    lists=["Arcane"],
    spells=[
        Spell(
            "Magic Missile",
            1,
            "One creature within \\rngmed range",
            """
            You send an unerring projectile made of pure magical energy towards the target.
            The target takes 1d6 energy damage.
        """,
            scaling="damage",
            tags=[],
        ),
        Spell(
            "Magic Missile Flurry",
            3,
            "Creatures in a \\areasmall radius within \\rngmed range",
            """
            You send an unerring projectile made of pure magical energy towards each target.
            Each target takes 1d8 energy damage.
        """,
            scaling="damage",
            tags=[],
        ),
        Spell(
            "Magic Missile Storm",
            6,
            "\\glossterm<Enemies> in a \\areamed radius within \\rnglong range",
            """
            You send an unerring projectile made of pure magical energy towards each target.
            Each target takes 2d8 energy damage.
        """,
            scaling="damage",
            tags=[],
        ),
        Spell(
            "Alter Magic Aura",
            1,
            "One Large or smaller magical object in \\rngmed range",
            """
            Make an attack vs. Mental against the target.
            \\hit One of the target's magic auras is altered (see \\pcref<Spellsense>).
            You can change the \\glossterm<ability tags> the aura has.
            In addition, you can decrease the \\glossterm<power> of the aura by up to half your power, or increase the power of the aura up to a maximum of your power.
            % No \\crit effect
        """,
            scaling="accuracy",
            tags=["Attune (self)"],
        ),
        Spell(
            "Suppress Item",
            1,
            "One Large or smaller magical object in \\rngmed range",
            """
            Make an attack vs. Mental against the target.
            \\hit All magical properties the target has are \\glossterm<suppressed>.
            % No relevant \\crit effect
        """,
            scaling="accuracy",
            tags=["Sustain (minor)"],
        ),
        Spell(
            "Dismissal",
            2,
            "One creature or object within \\rngmed range",
            """
            Make an attack against the target.
            If the target is an effect of an ongoing \\glossterm<magical> ability, such as a summoned monster or created object, the \\glossterm<difficulty rating> is equal to the \\glossterm<power> of the ability.
            Otherwise, this spell has no effect.
            \\hit The target is treated as if the ability that created it was \\glossterm<dismissed>.
            This usually causes the target to disappear.
            % No relevant \\crit effect
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Deattunement",
            3,
            "One creature within \\rngmed range",
            """
            Make an attack vs. Fortitude with a +2 bonus to \\glossterm<accuracy> against the target.
            \\hit The target stops being \\glossterm<attuned> to one effect of its choice that it is currently attuned to.
            % No \\glance effect
            \\crit The target stops being \\glossterm<attuned> to two abilities of its choice that it is currently attuned to.
            In addition, as a \\glossterm<condition>, it becomes unable to \\glossterm<attune> to any additional abilities.
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Reattunement",
            3,
            "Yourself or one \\glossterm<ally> within \\rngmed range",
            """
            The target can reattune to one ability that it stopped being attuned to since the start of the last round without spending an additional \\glossterm<attunement point>.
            Any choices and effects of the attuned ability are restored to their exact state before the attunement was broken.
        """,
            scaling="""
            \\rank<5> You can target an additional ally within range.
            \\rank<7> You can target an additional ally within range.
        """,
            tags=[],
        ),
        Spell(
            "Suppress Magic",
            2,
            "One magical effect within \\rngmed range (see text)",
            """
            The target must be the source or result of an active \\glossterm<magical> effect.
            This includes magical \\glossterm<conditions>, \\glossterm<sustained> effects, and \\glossterm<attuned> effects.
            It does not include \\glossterm<Curse> effects, which are more difficult to remove.
            This spell cannot be used to interrupt or negate immediate effects.
            Identifying non-visual magical effects to target with this spell may require the use of the Spellsense skill (see \\pcref<Spellsense>).

            Make an attack against the target.
            Its defense against this attack is equal to its \\glossterm<power>.
            \\hit The effect is \\glossterm<suppressed> as long as you \\glossterm<sustain> this ability.
        """,
            scaling="accuracy",
            tags=["Sustain (minor)"],
        ),
        Spell(
            "Dispel Magic",
            4,
            "One magical effect within \\rngmed range (see text)",
            """
            The target must be the source or result of an active \\glossterm<magical> effect with a duration.
            This includes magical \\glossterm<conditions>, \\glossterm<sustained> effects, and \\glossterm<attuned> effects.
            It does not include \\glossterm<Curse> effects, which are more difficult to remove.
            This spell cannot be used to interrupt or negate immediate effects.
            Identifying non-visual magical effects to target with this spell may require the use of the Spellsense skill (see \\pcref<Spellsense>).

            Make an attack against the target.
            Its defense against this attack is equal to its \\glossterm<power>.
            \\hit The effect is \\glossterm<dismissed>.
            If it required attunement, that attunement is broken.
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Malign Transferance",
            4,
            [
                "Yourself or an \\glossterm<ally> within \\rngmed range",
                "One other creature within that range",
            ],
            """
            The primary target must be currently affected by a \\glossterm<magical> \\glossterm<condition>.
            Make an attack vs. Mental against the secondary target.
            \\hit One magical condition of your choice is removed from the primary target and applied to the secondary target.
            % No \\glance effect; weird shenanigans if you autoremove the conditions
            \\crit As above, except that you can transfer any number of magical conditions in this way.
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Malign Confluence",
            7,
            [
                "Yourself and each \\glossterm<ally> within \\rngmed range",
                "One other creature within that range",
            ],
            """
            Make an attack vs. Mental against the secondary target.
            \\hit One magical condition of your choice is removed from each primary target and applied to the secondary target.
            % No \\glance effect; weird shenanigans if you autoremove the conditions
            \\crit As above, except that you can transfer any number of magical conditions from each primary target in this way.
        """,
            tags=[],
        ),
        Spell(
            "Spell Absoption",
            7,
            "Yourself",
            """
            Whenever you are targeted by a spell cast by another creature, before determining if it hits you, you absorb the spell.
            It has no effect on you.
            You cannot voluntarily allow spells to affect you while this effect lasts.
            After you absorb three spells in this way, this effect ends.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Enhance Magic",
            1,
            "Yourself",
            """
            You can cast this spell as a \\glossterm<minor action>.

            You gain a +2 \\glossterm<magic bonus> to your \\glossterm<magical> \\glossterm<power>.
        """,
            scaling="""
            \\rank<3> This spell can target an \\glossterm<ally> within \\rngmed range instead of you.
            \\rank<5> The bonus increases to +4.
            \\rank<7> The bonus increases to +8.
        """,
            tags=["Attune (self)"],
        ),
        # Is this worth the complexity it adds to the system?
        Spell(
            "Antimagic Field",
            7,
            "Special",
            """
            All other magical abilities and objects are \\glossterm<suppressed> within a \\areasmall radius \\glossterm<emanation> from you.
            % How much of this is redundant with suppression?
            Creatures within the area cannot activate, sustain, or dismiss magical abilities.
            % TODO: wording
            This does not affect aspects of creatures that cannot be suppressed, such as the knowledge of abilities.
            You cannot exclude yourself from this \\glossterm<emanation>.
            However, this spell does not prevent you from sustaining or dismissing this spell.
        """,
            tags=["Sustain (standard)"],
        ),
        Spell(
            "Dimensional Anchor",
            3,
            "One creature or object within \\rngmed range",
            """
            Make an attack vs. Mental with a +2 bonus to \\glossterm<accuracy> against the target.
            \\hit The target is unable to travel extradimensionally.
            This prevents all \\glossterm<Manifestation> effects and effects that teleport the target or move it between planes.
            \\glance As above, except that the effect ends at the end of the next round.
            % No relevant \\crit effect
        """,
            scaling="accuracy",
            tags=["Sustain (minor)"],
        ),
        Spell(
            "Dimensional Lock",
            5,
            None,
            """
            This spell creates a dimensional lock in a \\areamed radius \\glossterm<zone> from your location.
            Extraplanar travel into or out of the area is impossible.
            This prevents all \\glossterm<Manifestation> effects and effects teleport targets or move them between planes.
        """,
            scaling="""
            \\rank<7> The area increases to a \\arealarge radius \\glossterm<zone>.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Teleportation Ward",
            4,
            "Everything in a \\areamed radius \\glossterm<emanation> from you (see text)",
            """
            Teleportation into and out of the area is impossible.
            Any abilities which would cause creatures to teleport within the area have no effect.
        """,
            scaling="""
            \\rank<6> The area increases to a \\arealarge radius \\glossterm<emanation>.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Disrupt Casting",
            5,
            "One creature within \\rngmed range",
            """
            Make an attack vs. Mental against the target.
            \\hit As a \\glossterm<condition>, the target has a 50\\% chance to \\glossterm<miscast> any spell it casts.
            \\glance As above, except that the condition is removed at the end of the next round.
            \\crit As a \\glossterm<condition>, the target automatically \\glossterm<miscasts> any spell it casts.
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Absorb Magic",
            6,
            "Yourself",
            """
            You gain a +2 bonus to defenses against \\glossterm<magical> effects.
            In addition, the next time a \\glossterm<magical> attack beats your defenses, it has no effect on you.
            When you negate an attack in this way, this spell's effect ends.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Second Mind",
            4,
            "Yourself",
            """
            Choose a \\glossterm<magical> ability you are currently sustaining that requires either a \\glossterm<free action> or a \\glossterm<minor action> to sustain.
            That ability is automatically sustained as long as this effect lasts.
            This can allow you to sustain that ability for longer than 5 minutes.
            This does not allow you to make any choices as part of sustaining the chosen ability, such as telling summoned creatures how to act.
            Instead, you always make the same choice you made the last time you manually sustained the effect.
        """,
            scaling="""
            \\rank<6> Whenever you sustain an ability with this effect, you can also make any relevant choices as part of sustaining the ability.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Spellseal",
            2,
            "One creature within \\rngmed range",
            """
            Make an attack vs. Fortitude with a +2 bonus to \\glossterm<accuracy> against the target.
            \\hit As a \\glossterm<condition>, the next time the target tries to cast a spell, it \\glossterm<miscasts> that spell instead.
            When the target miscasts a spell, this effect ends.
            \\crit As above, except that the effect ends after the target miscasts two spells instead of only one.
        """,
            scaling="accuracy",
            tags=[],
        ),
        Spell(
            "Delay Teleportation",
            4,
            "See text",
            """
            You can cast this spell as a \\glossterm<minor action>.

            Whenever a creature or object would teleport into the area in a \\areamed radius \\glossterm<emanation> from you, that teleportation is delayed by a round.
            The teleporting creature or object remains stuck in the Astral Plane and can take no actions during that time.
            Creatures delayed in this way do not experience a delay, though they may be able to deduce that they were delayed based on observable evidence.

            Whenever something is delayed in this way, you learn its approximate size and location within the area, allowing you to know which space or spaces it will occupy when it arrives.
            Creatures and objects delayed by this effect remain delayed even if you move such that their destination is no longer within the area of this effect.
            This does not affect teleportation away from the area unless that teleportation's destination lies within the area.
        """,
            scaling="""
            \\rank<6> The delay increases to two rounds.
        """,
            tags=["Attune (self)"],
        ),
        Spell(
            "Spellward",
            4,
            "Creatures in a \\areamed radius \\glossterm<zone> within \\rngmed range",
            """
            You seal off an area's connection to magic.
            Whenever a creature casts a spell in the area, it has a 50\\% chance to \\glossterm<miscast> the spell.
        """,
            scaling="""
            \\rank<6> The area increases to a \\arealarge radius.
        """,
            tags=["Sustain (minor)"],
        ),
        Spell(
            "Cryptic Spells",
            2,
            "Yourself",
            """
            Whenever you cast a spell, you may choose a different spell you know.
            If you do, the visual effects and magical aura of the spell they are casting change to match your chosen spell.
            This affects inspection of the spell itself by any means, such as with the Spellsense skill (see \\pcref<Spellsense>).
            However, it does not alter the mechanical effects of the spell in any way.

            An observer can make a Spellsense check with a \\glossterm<difficulty rating> of 15 \\add your magical \\glossterm<power> to identify the spell's true nature.
            If the spell's effects depend on visual components, the spell may fail to work if the target alters the spell's visuals too much.
        """,
            scaling="""
            \\rank<4> You also gain a +1 \\glossterm<magic bonus> to \\glossterm<accuracy> with spells.
            \\rank<6> The accuracy bonus increases to +2.
        """,
            tags=["Attune (target)"],
        ),
    ],
    rituals=[
        Spell(
            "Dispel Curse",
            3,
            "Yourself or an \\glossterm<ally> within \\rngmed range",
            """
            All curses affecting the target are removed.
            This ritual cannot remove a curse that is part of the effect of an item the target has equipped.
            However, it can allow the target to remove any cursed items it has equipped.
        """,
            tags=[],
            ritual_time="24 hours",
        ),
    ],
    category="debuff, combat",
)

File: typescript/src/mystic_spheres/thaumaturgy.ts
import { MysticSphere } from '.';
import { BARRIER_COOLDOWN, CONDITION_CRIT } from './constants';

export const thaumaturgy: MysticSphere = {
  name: 'Thaumaturgy',
  shortDescription: 'Suppress and manipulate magical effects.',
  sources: ['arcane', 'domain'],

  cantrips: [
    {
      name: 'Detect Magic',

      effect: `
        Choose a \\arealarge cone from you.
        You know whether any \\magical abilities were activated in that area since the start of the last round.
        This does not provide any information about passive magical effects, such as attuned spells.
        It also does not provide any information about the number or location of those magical effects.
      `,
      roles: ['narrative'],
      scaling: {
        2: `
          You also learn if any passive magical effects existed in the area.
          This is separate from your knowledge of active magical effects.
        `,
        4: 'You can choose to create a \\hugearea cone instead.',
        6: `You also learn the number of magical effects in the area.`,
      },
      tags: ['Detection'],
    },
  ],
  spells: [
    {
      name: 'Magic Missile',

      // -1 range in exchange for the strong miss effect
      attack: {
        hit: `
          \\damagerankone.
        `,
        missGlance: true,
        targeting: `
          Make an attack vs. Armor against something within \\shortrange.
          This attack ignores \\glossterm{cover} and all \\glossterm{miss chances}.
        `,
      },
      narrative: `
        An unerring projectile made of pure magical energy streaks towards your foe, inflicting damage unavoidably.
      `,
      rank: 1,
      roles: ['burst'],
      scaling: 'damage',
    },

    {
      name: 'Distant Magic Missile',

      attack: {
        hit: `
          \\damagerankfive, and any \\glossterm{extra damage} is doubled.
        `,
        missGlance: true,
        targeting: `
          Make an attack vs. Armor against something within \\longrange.
          This attack ignores \\glossterm{cover} and all \\glossterm{miss chances}.
        `,
      },
      narrative: `
        An unerring projectile made of pure magical energy streaks towards your foe, inflicting severe damage unavoidably.
      `,
      rank: 7,
      roles: ['burst'],
      scaling: 'damage',
    },

    {
      name: 'Mighty Magic Missile',

      // -1 range for miss effect and cover/concealment effect
      attack: {
        hit: `
          \\damagerankfour, and any \\glossterm{extra damage} is doubled.
        `,
        missGlance: true,
        targeting: `
          Make an attack vs. Armor against something within \\shortrange.
          This attack ignores \\glossterm{cover} and all \\glossterm{miss chances}.
        `,
      },
      narrative: `
        An unerring projectile made of pure magical energy streaks towards your foe, inflicting massive damage unavoidably.
      `,
      rank: 4,
      roles: ['burst'],
      scaling: 'damage',
    },

    // treat rank as 1 lower for determining area, but not damage. Ignoring cover and miss
    // chances is less exciting for AOE than single target, since single target also gets
    // the half damage on miss effect.
    {
      name: 'Magic Missile Storm',

      attack: {
        hit: `
          \\damagerankfour.
        `,
        missGlance: true,
        targeting: `
          Make an attack vs. Armor against all \\glossterm{enemies} within a \\largearea radius from you.
          This attack ignores \\glossterm{cover} and all \\glossterm{miss chances}.
        `,
      },
      narrative: `
        A barrage of unerring projectiles made of pure magical energy streak towards an area, inflicting damage unavoidably.
      `,
      rank: 6,
      roles: ['clear'],
      scaling: 'damage',
    },

    {
      name: 'Suppress Item',

      attack: {
        hit: `All magical properties the target has are \\glossterm{suppressed}.`,
        targeting: `
          Make an attack vs. Mental with a +4 \\glossterm{accuracy} bonus against one Large or smaller \\magical object within \\medrange.
          If the object is attended by a creature, the attack must also beat the attending creature's Mental defense.
        `,
      },
      rank: 2,
      roles: ['softener'],
      scaling: 'accuracy',
      type: 'Sustain (minor)',
    },

    {
      name: 'Dismissal',

      attack: {
        hit: `The target is treated as if the ability that created it was \\glossterm{dismissed}.
        This usually causes the target to disappear.`,
        targeting: `
        Make an attack against something within \\medrange.
        If the target is an effect of an ongoing \\magical ability, such as a summoned monster or created object, its defense against this attack is equal to 5 + twice the \\glossterm{rank} of the effect.
        Otherwise, this spell has no effect.
        `,
      },

      rank: 2,
      roles: ['softener'],
      scaling: 'accuracy',
    },

    {
      name: 'Suppress Magic',

      attack: {
        hit: `The effect is \\glossterm{suppressed} as long as you \\glossterm{sustain} this ability.`,
        targeting: `
          Choose a single 5-ft. square within \\medrange.
          Make an attack against a random \\magical effect active in that square, if any exist.
          This can only target magical effects that are sustained or which have a specific duration.
          This includes brief effects, conditions, and \\abilitytag{Sustain} abilities, but it does not include attuned effects or permanent magical abilities on creatures.
          It also does not include \\abilitytag{Curse} effects, which are more difficult to remove.
          The attack cannot \\glossterm{explode} by any means.

          The target's defense against this attack is equal to 5 \\add twice its \\glossterm{rank}.
          For effects that have no specific rank, such as some monster abilities, treat their rank as being equal to one third of their level.

          This spell cannot be used to interrupt or negate immediate effects, such as spells being cast.
          Identifying non-visual magical effects can be difficult, so you may have to guess which area to target.
        `,
      },

      rank: 2,
      roles: ['softener'],
      scaling: 'accuracy',
      type: 'Sustain (minor)',
    },

    {
      name: 'Greater Suppress Magic',

      functionsLike: {
        name: 'suppress magic',
        exceptThat: `
          it can also affect \\abilitytag{Curse} effects.
        `,
      },
      rank: 5,
      roles: ['softener'],
      scaling: 'accuracy',
      type: 'Sustain (minor)',
    },

    {
      name: 'Dispel Magic',

      functionsLike: {
        exceptThat: `
          on a hit, the effect ends completely instead of being suppressed.
        `,
        name: 'suppress magic',
      },
      rank: 4,
      roles: ['softener'],
      scaling: 'accuracy',
    },

    {
      name: 'Imbue Weapon',

      effect: `
        Choose one weapon you are touching and proficient with.
        You can use the higher of your \\glossterm{magical power} and \\glossterm{mundane power} to determine your damage with \\glossterm{strikes} using that weapon (see \\pcref{Power}).
      `,
      rank: 1,
      roles: ['attune'],
      type: 'Attune',
    },

    {
      name: 'Spellvoid',

      effect: `
        Whenever you are targeted by a spell cast by another creature, before determining if it hits you, you absorb the spell.
        It has no effect on you.
        This does not prevent the spell from affecting other creatures.
        You cannot voluntarily allow spells cast by other creatures to affect you while this effect lasts.
        After you absorb three spells in this way, this ability is \\glossterm{dismissed}.
      `,
      rank: 7,
      roles: ['attune'],
      type: 'Attune (deep)',
    },

    {
      name: 'Enhance Magic -- Might',

      effect: `
        Whenever you cast a spell that does not have the \\abilitytag{Sustain} or \\abilitytag{Attune} tags, you can activate this effect as a \\glossterm{minor action}.
        If you do, the spell deals 1d6 \\glossterm{extra damage} when it deals damage for the first time.
        After you enhance a spell in this way, this ability is \\glossterm{dismissed}.
      `,
      rank: 1,
      roles: ['attune'],
      scaling: {
        3: `The extra damage increases to 2d6.`,
        5: `The extra damage increases to 2d10.`,
        7: `The extra damage increases to 4d10.`,
      },
      type: 'Attune',
    },

    {
      name: 'Enhance Magic -- Precision',

      effect: `
        Whenever you cast a spell that does not have the \\abilitytag{Sustain} or \\abilitytag{Attune} tags, you can activate this effect as a \\glossterm{minor action}.
        If you do, you become \\focused and \\honed for the rest of that round.
        After you enhance a spell in this way, this ability is \\glossterm{dismissed}.
      `,
      rank: 1,
      roles: ['attune'],
      type: 'Attune',
    },

    {
      name: 'Enhance Magic -- Echo',

      effect: `
        Whenever you cast a spell that does not have the \\abilitytag{Sustain} or \\abilitytag{Attune} tags, you can activate this effect as a \\glossterm{minor action}.
        If you do, the spell \\glossterm{repeats} during your next action.
        After you enhance a spell in this way, this ability is \\glossterm{dismissed}.
      `,
      rank: 5,
      roles: ['attune'],
      type: 'Attune',
    },

    {
      name: 'Dimensional Seal',

      attack: {
        hit: `
          The target cannot be \\glossterm{teleported}.
          An object affected by this spell is left behind if it is carried by a creature that teleports.
        `,
        targeting: `
          Make an attack vs. Mental with a +2 accuracy bonus against something within \\medrange.
        `,
      },
      rank: 2,
      roles: ['narrative'],
      scaling: 'accuracy',
      type: 'Sustain (attuneable, minor)',
    },

    {
      name: 'Dimensional Lock',

      effect: `
        This spell creates a dimensional lock in a \\largearea radius \\glossterm{zone} from your location.
        Extraplanar travel into or out of the area is impossible.
        This prevents all \\abilitytag{Manifestation} effects and effects teleport targets or move them between planes.
      `,
      rank: 4,
      roles: ['narrative'],
      type: 'Sustain (attuneable, minor)',
    },

    {
      name: 'Absorb Magic',

      effect: `
        You gain a +1 \\glossterm{enhancement bonus} to your defenses against \\magical effects.
        In addition, the next time a \\magical attack beats your defenses, it has no effect on you.
        After you negate two attacks in this way, this spell's effect ends.
      `,
      rank: 6,
      roles: ['attune'],
      type: 'Attune (deep)',
    },

    {
      name: 'Second Mind',

      effect: `
        Choose a \\magical ability you are currently sustaining that requires either a \\glossterm{free action} or a \\glossterm{minor action} to sustain.
        That ability is automatically sustained as long as this effect lasts.
        % Divine Offering breaks permanent sustain; there are probably others
        This can allow you to sustain that ability for longer than 5 minutes.
        However, the ability automatically ends after 10 minutes, ending your attunement to this spell.

        Automatically sustaining an ability with this spell does not allow you to make any new choices, such as giving summoned creatures new instructions.
        Instead, you always make the same choice you made the last time you manually sustained the effect.
        You can still manually sustain the ability to change your choices while this spell is active.
      `,
      rank: 4,
      roles: ['attune'],
      type: 'Attune',
    },

    {
      name: 'Spellseal',

      // Action skip is 2 EA and must be combined with limited scope, so we get r3. Drop
      // to r1 because it only works on spellcasters.
      attack: {
        hit: `
          The target's magic is partially sealed as a \\glossterm{condition}.
          The first time it tries to cast a spell, the spell automatically fails with no effect instead.
          When the target fails to cast a spell in this way, this ability is \\glossterm{dismissed}, and the target becomes immune to this spell until it finishes a \\glossterm{short rest}.
        `,
        targeting: `
          Make an attack vs. Fortitude against one creature within \\medrange.
        `,
      },
      rank: 1,
      roles: ['stasis'],
      scaling: 'accuracy',
    },

    {
      name: 'Delay Teleportation',

      effect: `
        Whenever a creature or object would teleport into a \\hugearea radius \\glossterm{emanation} from you from outside of that area, that teleportation is delayed by a round.
        The teleporting creature or object remains stuck in the Astral Plane and can take no actions during that time.
        Creatures delayed in this way do not experience a delay, though they may be able to deduce that they were delayed based on observable evidence.

        Whenever something is delayed in this way, you learn its approximate size and location within the area, allowing you to know which space or spaces it will occupy when it arrives.
        Creatures and objects delayed by this effect remain delayed even if you move such that their destination is no longer within the area of this effect.
        This does not affect teleportation away from or within the area.
      `,
      rank: 5,
      roles: ['attune'],
      type: 'Attune',
    },

    {
      name: 'Cryptic Spells',

      effect: `
        Whenever you cast a spell, you may choose to mimic the identifying characteristics of a different spell you know.
        If you do, the \\glossterm{verbal components}, \\glossterm{somatic components}, visual effects, and magical aura of the spell you are casting change to match the mimic spell.
        This affects inspection of the spell itself by any means.
        However, it does not alter the mechanical effects of the spell in any way.
        This change cannot fully remove verbal, somatic, or visual components.
        If the mimic spell does not have an identifying characteristic, the original characteristic of the spell you are casting is unchanged.

        An observer who gets a \\glossterm{critical success} to identify the spell's effects can identify the true spell that you are casting (see \\pcref{Identify Magical Effect}).
      `,
      rank: 3,
      roles: ['narrative'],
      type: 'Attune',
    },

    {
      name: 'Reflect Magic',

      effect: `
        You are \\braced this round.
        In addition, whenever a creature within \\medrange of you misses or \\glossterm{glances} you with a \\magical attack this round, that creature treats itself as a target of that attack in addition to any other targets.
        The attacker cannot choose to reduce its accuracy or damage against itself.
      `,
      rank: 2,
      roles: ['turtle'],
      tags: ['Swift'],
    },

    {
      name: 'Greater Reflect Magic',

      effect: `
        You are \\braced this round.
        In addition, whenever a creature within \\medrange of you makes a \\magical attack against you this round, that creature treats itself as a target of that attack in addition to any other targets.
        The attacker cannot choose to reduce its accuracy or damage against itself.
      `,
      rank: 5,
      roles: ['turtle'],
      tags: ['Swift'],
    },

    {
      name: 'Wall of Magic Impedance',

      cost: BARRIER_COOLDOWN,
      effect: `
        You create a \\medarealong \\glossterm{wall} of magical energy within \\medrange.
        The wall is visible as a shimmering magical membrane that does not block sight.
        It has \\glossterm{hit points} equal to three times your \\glossterm{power}, and is destroyed when its hit points become negative.
        Objects, creatures, and \\glossterm{mundane} abilities can pass through the wall freely, but any \\magical ability treats the wall as an impassable barrier.
      `,
      rank: 5,
      roles: ['hazard'],
      scaling: {
        7: "The wall's hit points increase to four times your power.",
      },
      tags: ['Barrier', 'Manifestation'],
      type: 'Sustain (attuneable, minor)',
    },

    {
      name: 'Font of Power',

      // Treat "yourself and adjacent allies who used a magical ability" as being roughly
      // "any two", so 0.5 EA. That requires a -2dr drop.
      // r0 area gives drX+1, drop to drX-1 for buff effect.
      attack: {
        hit: `
          \\damagerankone.
        `,
        missGlance: true,
        targeting: `
          Make an attack vs. Reflex against all \\glossterm{enemies} adjacent to you.
          Then, you and all adjacent \\glossterm{allies} who used a \magical ability this round are \\glossterm{briefly} \\empowered.
        `,
      },
      rank: 2,
      roles: ['generator'],
      scaling: 'damage',
      tags: ['Swift'],
    },

    {
      name: 'Mighty Font of Power',

      // Treat "yourself and adjacent allies who used a magical ability" as being slightly
      // weaker than "any two", so 0.4 EA. That requires a -2dr drop.
      // r0 area gives drX+1, drop to drX-1 for buff effect.
      functionsLike: {
        name: 'font of power',
        exceptThat:
          'the damage increases to \\damagerankfive, and any \\glossterm{extra damage} is doubled.',
      },
      rank: 6,
      roles: ['generator'],
      scaling: 'damage',
      tags: ['Swift'],
    },
  ],
};
